django_socio_grpc/protobuf/generators.py

```diff
@@ -144,6 +181,31 @@ def _generate_message(self, message: ProtoMessage):
         self._writer.write_line("}")
         self._writer.write_line("")
 
+    def _generate_enum(self, enum: Enum, indices: dict[str, int]):
```

Comment:
why not creating a generation plugin used by default ? 
As our goal is to rawork the all generators class as plugin to have multiple generation source (drf serializer, vs pydantic models or others) this new feature should be directly a generation plugin to avoid making the next work more difficult.

Reply:
didn't know about generation plugins and the current interest to use them. Will refactor if that's okay for @legau

Reply:
so after looking a bit into it, i don't really understand how that would be applicable.
What i understood is that GenerationPlugin are used to modify the request or response message of a ProtoRPC.
In the generateproto command, all generation plugins are run when the service are registered, and then the RegistryToProtoGenerator use the registry to write the actual .proto file.
Now what i don't understand, is that i need to write lines specifically for the enums, in addition to the field of enum type in the message, how does it relates to using a GenerationPlugin ? I can see that to generate services and messages the same method is used.

Reply:
Maybe is not really applicable until we have the all generation system as plugin.

But in my mind:
- In your EnumGenerationPlugin you iterate over the fields. 
- If the field is a textchoice, enum or anything that create a enum you call the _from_choice_field method directly in the plugin.
- This register the enum in the protoMessage.
- This is taken into account when writing file

As the enums is your version are not encapsulated in the message but directly in the global scope maybe it will be necessary to add a new params to run_validation_and_transform named shared_message or something similar that will be completed with the enum created in the plugin.

This way the user could choose different way to register the enum depending of the plugin he choose.

Reply:
the _from_choice_field only return the ProtoField with the right field_type (string, int32, or an Enum) so it's already called before the plugins.
Currently we don't have **any** information to know where to write the enum.

Maybe an idea would be to add a "enums" field to ProtoMessage which should be initialized by a GenerationPlugin, and that field would contain informations to tell where to output it like :
- Global scope
- Global scope inside a message
- Directly in the message

At the start of the generation we go over all message, here we could do some process for all the cases, but still we need to implement code to handle all those cases when writing the .proto

That would still be a very limited solution imo, we need to have a way to structure the data in a more complex way than ProtoService -> ProtoRPC -> ProtoMessage to make this really modular no ?

Reply:
for your first sentence: The main idea is to not deal with enum before calling plugin but only in plugin to split the code logic into plugin if possible.

To the other thoughts:
I see what you mean, the write process to the proto file is not configurable. Only the data used to process is.

I think you may be right but no need to change the all structure of the data. But yes complexy it a little bit
Meaning you still have ProtoService -> ProtoRPC -> ProtoMessage and in ProtoMessage you add a new attributes named proto_extra_tools that is also a class (ProtoExtraTools) with for now only a enums key that describe the enum

Then in AppHandlerRegistry adding the same attribute named proto_extra_tools that also is an instance of ProtoExtraTools. And this is this one that is populated and read when adding import or global Enum.

Reply:
IMO the scope depend of where you store it.

If you store it in proto_extra_tools of ProtoMessage -> MyEnum (innermost scope)
If you stock it in proto_extra_tools of AppHandlerRegistry -> .MyEnum (outermost scope)
If you stock it in proto_extra_tools of ProtoMessage of a new message that have the name of the enum -> .MyEnum.Enum (outermost scope + wrapped in message)

If your question was more about how to dynmically change the fied_type of the enum field this is done in the plugin.
Before plugin the field is a simple string as before your implementation.
Then it pass into the plugin. 
The plugin iterate over the field and find the enums one.
It store the enum into proto_extra_tools (of the ProtoMessage or the app_handler_registry (we may need to pass the instance of the AppHandlerRegistry to the plugins) depending of the pugin you choose tu use)

Reply:
I used the property setter and it works great, before cleaning my code i have a last question.
I put the EnumGenerationPlugin as a  default plugin in the django settings, but it raise a circle dependency error.
How should it be done ? 

To test, I added `DEFAULT_GENERATION_PLUGINS` to the `IMPORT_STRING` list and added some code to convert the class to an instance when `DEFAULT_GENERATION_PLUGINS`  is retrieved, not sure it is expected to do that.

Reply:
yes, we need to pass an instance, but we can't import generation plugins in the settings as it would raise a circular import

Reply:
Do you know which import is circular in the first time ?
Then you have two choices:
- Splitting the file to isolate the circular import in an other file if not needed 
- Importing it inside the plugin and not at the top of the file

Reply:
thanks, did not know about directly importing inside a method. 
Still have to write the doc and solve one problem and it should be good to review tomorrow